rfc-163/relocating-logic-from-cdn.md

```diff
@@ -0,0 +1,107 @@
+# Appendix: Proposal for relocating logic from the CDN layer to other parts of the stack
+
+Things that could be moved to WAF:
+
+- IP allowlisting on staging and production EKS[^ip-allowlist]
+- Requiring HTTP Basic auth on integration[^http-basic-1][^http-basic-2] (unless the user's IP is in the allowlist[^http-basic-allowlist])
+- IP denylisting[^ip-denylist]
+  - This functionality is currently unused (the dictionary that the denylist is read from is empty), but it exists in case we ever need to quickly block IP addresses (for example, during an incident).
```

Comment:
I think there's a problem with these three around caching.

If the CDN has a page in its cache, then it won't make a request to AWS (where the WAF lives), so the IP allowlisting decision / basic auth / denylisting decision happens to late. For example, if a user with a banned IP requests a page that's cached, how is the CDN to know it shouldn't serve that to them?

We could work around this by disabling caching (probably not acceptable, even in non-prod environments as it would be a big disparity between the environments), or maybe by [doing something clever with cookies and the Vary header](https://www.fastly.com/blog/best-practices-using-vary-header#vary-cookie) (but that might end up being more complex / confusing than the current situation).

I think this is worth a bit of a discussion, because it's possibly a showstopper. It's quite possible I've got the wrong end of the stick though.

Reply:
@ChrisBAshton and I had a chat about this earlier and we actually think it's probably ok for denylisting to happen in the WAF.

If a denylisted user makes a request of any kind, they will be blocked by the WAF and get an HTTP 403 in response. Origin won't be hit, and client errors aren't (or at least, shouldn't be) cached by the CDN.

If a non-denylisted user makes a cacheable request (e.g. a `GET` to the homepage), they will get a 200 and this will be cached. If the denylisted user then makes the same request, they will receive the cached version (and again, origin won't be hit).

We think that's OK, because to our knowledge denylisting is intended to protect origin from attacks, rather than for access control.

Basic auth will still have to live in the CDN for the reason you mentioned (otherwise we'd have to add `Vary: Authorization` to the response, which would pretty much defeat caching).

What do you think @richardTowers ?

Reply:
> denylisting is intended to protect origin from attacks, rather than for access control

It's true that it's not used for access control, but I'm not sure if that means it's just there to protect origin. It also gives us some useful protection at the CDN level. For example, if someone's downloading terrabytes of assets from one IP, Fastly might ask us if we could make them stop doing that (or if Fastly ever move us to usage based billing, we might want to make them stop), and I think IP denylisting is the only way we could do that at the moment.

> client errors aren't (or at least, shouldn't be) cached by the CDN

Definitely [citation needed] for this - I seem to remember 404s being cached (which is a reasonable choice for a CDN to make - there wasn't a resource here last time it was requested, there won't be a resource here this time). I'm not sure whether 403s get different treatment. We could fix that in the VCL for sure, if we haven't already (but then we're adding more code to the VCL to enable us to remove some code from the VCL, so we'd have to consider the cost / benefit of that).

> otherwise we'd have to add Vary: Authorization to the response, which would pretty much defeat caching

🤔  I wonder how bad this would be actually. Browsers don't do any kind of nonce / salting of Authorization headers, so they're always the same for the same username / password. Our basic auth use cases don't have a massive variety of username / password combinations, so I think we'd get a reasonable cache hit rate for them. Maybe that one could move to the WAF.

Reply:
https://www.rfc-editor.org/rfc/rfc7231#section-6.5.4 lists a few 4xx responses as cacheable by default, but notably not 403s (404s are cacheable, as are 405s, 410s, and a few others). Whether Varnish implements this is another question...

Reply:
https://github.com/varnish/Varnish-Book/blob/master/varnish_book.rst#the-initial-value-of-beresp-ttl

> Only the following status codes will be cached by default:

> 200: OK
> 203: Non-Authoritative Information
> 300: Multiple Choices
> 301: Moved Permanently
> 302: Moved Temporarily
> 304: Not modified
> 307: Temporary Redirect
> 410: Gone
> 404: Not Found

So not quite the same as RFC7231, but it won't cache 403s so you can ignore that argument :)

Reply:
> For example, if someone's downloading terrabytes of assets from one IP, Fastly might ask us if we could make them stop doing that (or if Fastly ever move us to usage based billing, we might want to make them stop), and I think IP denylisting is the only way we could do that at the moment.

I think we'd be more likely to use JA3 denylisting for this(?)

I'm actually not sure how valuable IP denylisting is as a general concept, given how easy it is for an attacker to get access to a bunch of IPv4 addresses, and given that IPv6 is a thing, but I'm open to being persuaded 😅

> > otherwise we'd have to add Vary: Authorization to the response, which would pretty much defeat caching
> 
> 🤔 I wonder how bad this would be actually. Browsers don't do any kind of nonce / salting of Authorization headers, so they're always the same for the same username / password. Our basic auth use cases don't have a massive variety of username / password combinations, so I think we'd get a reasonable cache hit rate for them. Maybe that one could move to the WAF.

Unfortunately I think it'd have to be more complicated than that, because we allow access to integration if _either_ the user provided the password _or_ their IP is in the allowlist. We'd have to do something like either:

- Set `Vary: Fastly-Client-IP, Authorization` on the response (which would essentially defeat caching), or
- Set `Vary: Authorization` always, and `Vary: Fastly-Client-IP` only if the `Authorization` header is missing (which actually might be a good shout, depending on how complex it'd be to implement in WAF 🤔)

I think I might tentatively move that back to the "move to WAF" section, with a caveat that it might prove to be too complicated.

Reply:
Thanks, that's useful context! I think we'll need to leave the denylist in the CDN, in that case (and see what we can do about ensuring parity between Fastly and Cloudfront)